pkg/action/pull.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package action
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/pflag"
+
+	"k8s.io/helm/pkg/chartutil"
+	"k8s.io/helm/pkg/cli"
+	"k8s.io/helm/pkg/downloader"
+	"k8s.io/helm/pkg/getter"
+	"k8s.io/helm/pkg/repo"
+)
+
+// Pull is the action for checking a given release's information.
+//
+// It provides the implementation of 'helm pull'.
+type Pull struct {
+	ChartPathOptions
+
+	Out      io.Writer       // TODO: refactor this out of pkg/action
+	Settings cli.EnvSettings // TODO: refactor this out of pkg/action
+
+	Devel       bool
+	Untar       bool
+	VerifyLater bool
+	UntarDir    string
+	DestDir     string
+}
+
+// NewPull creates a new Pull object with the given configuration.
+func NewPull() *Pull {
+	return &Pull{}
+}
+
+// Run executes 'helm pull' against the given release.
+func (p *Pull) Run(chartRef string) error {
+	c := downloader.ChartDownloader{
+		HelmHome: p.Settings.Home,
+		Out:      p.Out,
+		Keyring:  p.Keyring,
+		Verify:   downloader.VerifyNever,
+		Getters:  getter.All(p.Settings),
+		Username: p.Username,
+		Password: p.Password,
+	}
+
+	if p.Verify {
+		c.Verify = downloader.VerifyAlways
+	} else if p.VerifyLater {
+		c.Verify = downloader.VerifyLater
+	}
+
+	// If untar is set, we fetch to a tempdir, then untar and copy after
+	// verification.
+	dest := p.DestDir
+	if p.Untar {
+		var err error
+		dest, err = ioutil.TempDir("", "helm-")
+		if err != nil {
+			return errors.Wrap(err, "failed to untar")
+		}
+		defer os.RemoveAll(dest)
+	}
+
+	if p.RepoURL != "" {
+		chartURL, err := repo.FindChartInAuthRepoURL(p.RepoURL, p.Username, p.Password, chartRef, p.Version, p.CertFile, p.KeyFile, p.CaFile, getter.All(p.Settings))
+		if err != nil {
+			return err
+		}
+		chartRef = chartURL
+	}
+
+	saved, v, err := c.DownloadTo(chartRef, p.Version, dest)
+	if err != nil {
+		return err
+	}
+
+	if p.Verify {
+		fmt.Fprintf(p.Out, "Verification: %v\n", v)
```

Comment:
Look into refactoring out any direct output from the `action` package

Reply:
done